### ISSUE TITLE
Preserve JSDocs in *.d.ts files when stripping comments

### ISSUE THREAD

**Issue by dotnetprofessional** [Microsoft/TypeScript#14619]
# Suggestion/Feature
When using the switch ```removeComments``` it will strip out all comments including the ones you may consider useful to the end consumer such as JSDocs. I'd like to strip out any internal comments but preserve any JSDocs comments. It seems that an option to preserve comments in types would be sufficient.

## Existing functionality
```json
"removeComments": true,
"declaration": true,
```
removes all comments including those in *.d.ts files.

## Suggested functionality
```json
"removeComments": true,
"declaration": true,
```
I would think by default you'd want to include the JSDocs in your type defs, so this would remove all comments from the code, but retain any comments associated with the types ie JSDocs.

If however, there is a need to remove all comments then an additional switch may be required to serve that purpose:

```json
"removeCommentsIncludingDeclarations": true,
"declaration": true,
```
Would produce the same result we have today.

**Comment by pocesar** [comment]
it makes no sense to remove the comments from declaration files, since they are mainly used during development and won't go into production, I was expecting too that `removeComments` would only remove them from transpiled .js files.

**Comment by DanielRosenwasser** [comment]
Legitimate question - why is anyone using `removeComments`? I would assume that you're using a minifier if you actually care about removing comments, but maybe I'm not aware of other scenarios.

**Comment by pocesar** [comment]
unminified code not for the browser, mainly. split code base into composable components that will later become part of something bigger (but while in dev, useful being on the consumer part of it). comments in JS = not useful. Comments in .d.ts, uber useful.

**Comment by patsissons** [comment]
Is this still on the horizon? This is a bit of a pain point with respect to my work-around for #21566 when generating declaration files. Fortunately in my case the declaration file is equivalent to the input source file so i can just replace the generated declaration file with a copied and renamed source file in automation.

**Comment by HolgerJeromin** [comment]
UglifyJS strips all comments but those with @license or @preserve.

**Comment by alexsasharegan** [comment]
@DanielRosenwasser, I frequently publish bite-size packages using typescript as my only build step. I usually want my code minified, but my comments available in the declarations since they are doc comments. This is a wonderfully simple way for my to publish packages rather than incorporating a bundler with plugins.

**Comment by dotnetprofessional** [comment]
HAPPY ONE YEAR ANNIVERSERY!!!!!

Its been exactly 1 YEAR today since this was opened. Are we any closer to having this resolved? This doesn't seem to be a lot of work to fix, but given we've had two duplicates since this was opened, it is something people want. I know I'd like to use it on my OSS projects...

**Comment by mindplay-dk** [comment]
On a related note: if/when this does get fixed, it would be nice to also adjust the .d.ts output slightly, so that there's a space between doc-blocks and declarations. Currently it's all crammed together.

**Comment by dotnetprofessional** [comment]
@RyanCavanaugh why mark my reply with a thumbs down? I think its quite valid to ask what's going on with a feature that's been 'In Discussion' for over a year. It would have been more helpful to the community to have spent time replying with an actual update, rather than just being rude!

**Comment by championswimmer** [comment]
I'd suggest
`comments: "never"` = no comments
`comments: "declaration"` = comments are put into the .d.ts files
`comments: "emit"` = comments are put into emitted js files

**Comment by bluelovers** [comment]
hope can have this, 
i don't think we should use another tool when we wanna remove comment on .js, but keep it on .d.ts

**Comment by Ionaru** [comment]
@DanielRosenwasser to answer your question:
I think there's a difference between comments used for development and comments used for API consumers.

Comments for development (Starting with `//` or JSDoc with `@internal`) live only in the source code and are to help future development easier. These commends do not need to be in the distribution files as they are of no use for the consumer.

Comments for API consumers (JSDoc) should live either in the declaration `d.ts` file or in the compiled distribution files. These comments provide explanations of functions, parameters and other things a consumer of the API can use.

My ideal situation:
* Regular comments used for development and JSDocs with `@internal` are not outputted to the disctribution files.
* The docstring or JSDoc is outputted to the `d.ts` file.
* The docstring or JSDoc is optionally outputted to the `.js` file.

Since VSCode looks at the `d.ts` files for documentation, emitting the JSDoc to `d.ts` files should have priority.

This could use the tsconfig suggestion @championswimmer made.

**Comment by bbugh** [comment]
I was looking at releasing a typescript library and was really surprised when my comments were either being stripped altogether or completely left in. Having intellisense would be really helpful to end users, but a lot of comments just makes the package size larger. 

Though, it did help explain why so many of my third party libraries have nice types but no intellisense documentation. 

Is there anything we can do as a community to move this forward? I see official comments, but no confirmation about how to resolve. Given that the proposed solutions are bigger (maybe) not backwards-compatible changes, maybe getting buy-in from one of the core team is important before starting work?

**Comment by cinderblock** [comment]
### Question

Should JSDoc comments be stripped from public APIs of compiled output `*.js` files?

----

While I'd be happy with @championswimmer's suggestion, I don't see that as compatible with existing options. Based on one discussion that I've had with some on the TS team, one high priority is keep consistent non confusing compiler options. Therefore I'm betting that adding a `comments` option alongside the existing boolean `removeComments` option is a non-starter.

### Backwards Compatible Solution

What if `removeComments` was expanded from `boolean` to `boolean | 'none' | 'code' | 'internal' | 'all'`?

```js
if (removeComments === false) removeComments = 'none';
if (removeComments === true) removeComments = 'code';
```

 - `none` - Preserve all comments
 - `code` - Preserve copyright comments
 - `internal` - Preserve JSDoc and copyright comments
 - `all` - Remove **all** comments

----

### Alternative

As the `removeComments` option does not remove `/*!` comment blocks, we could add an extra option like `preserveJSDocComments` that would preserve the appropriate comments when `removeComments` is set to `true`.

**Comment by loganpowell** [comment]
I was going to submit a separate issue, but I believe mine is related. Why can't TypeScript give us docstrings?

**Comment by nealeu** [comment]
HAPPY TWO YEAR ANNIVERSARY.

@RyanCavanaugh (or anyone else at Microsoft): are you going to respond or just :-1: this too?

Looking at recent releases I'm getting tempted to look at Kotlin for JS now.  Building and using libraries in with Typescript is still way off from the simplicity and consistency of Java/Maven ecosystem.
Now that MS has control of NPM too, then isn't this an opportunity to bring the JS ecosystem up.

But until then... can we please have this simple feature to retain JSDoc... PLEASE!

**Comment by bluelovers** [comment]
hope one day something like this can be merge
https://github.com/microsoft/TypeScript/pull/31813

**Comment by rbonestell** [comment]
[Official TypeScript documentation](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html) describes the implementation of JSDoc without mentioning this behavior. I think the implicit behavior should be something such as the solutions described above, but the documentation should _explicitly_ describe the behavior.

**Comment by mindplay-dk** [comment]
I'm sure a pull request (with test-coverage and documentation) wouldn't be rejected by the team - this is one of those "duh, obvious" features no one can really object to.

But to those of you who think this is a slam dunk and why doesn't it "just" get done: I gave it a shot, and about an hour later, all I've managed to do is run the test-suite, ha ha.

The feature itself is probably something like changing the type of `removeComments` to something like `true | false | "preserve-doc-blocks"`, then testing for that in the condition [here](https://github.com/microsoft/TypeScript/blob/4463194ddadfdeb54aa0d642343634bec1d5dc47/src/compiler/emitter.ts#L1208) - but also, the test-suite needs to be updated, and doesn't seem to have test-coverage for removal of comments in the first place, so that probably needs to be added first, the feature needs documentation, and so on.

Iterating on a codebase of this size and complexity is never quick or easy - there are only so many developers working on this, and I'm sure they have priorities, and yeah, sometimes apparently simple things take more than two years. If you think it's so easy, why don't you do it and submit a PR 😀

**Comment by Zzzen** [comment]
Currently, all comments are removed from d.ts when `removeComments ` is false. Am I missing anything?

![image](https://user-images.githubusercontent.com/6630042/101135883-6d581880-3647-11eb-86d3-3c44167c6ea0.png)


[playground](https://www.typescriptlang.org/play?#code/PTAEEsDsBcFMCcBmBDAxrUAxA9tgULAB4AO280EMCK6WuoA3nqKCKNABbgDOEvyobtHhQA5s1CJwsADYATAIIAuQcLEBuPAF88eNpx59QA1DOTduBEmQqnzvAELJ4jCfq69DAyAFcAtgBGCBJSsnIOKr6BCKAAvKAADJo6QA)

**Comment by thealjey** [comment]
@Zzzen yes, you are
this is about JSDoc comments specifically

![image](https://user-images.githubusercontent.com/2216553/101139384-c442fb00-3611-11eb-80fc-06c8335852c2.png)

**Comment by Zzzen** [comment]
😕  Do you mean normal comments are never preserved by design?

**Comment by thealjey** [comment]
in a declaration file?
of course they are not

**Comment by Zzzen** [comment]
So do you have any idea why `// interface Foo` is removed in `.D.TS` section? @thealjey

**Comment by cinderblock** [comment]
@Zzzen, because "`// ...`" is not a valid way for JSDoc to begin.

JSDoc *must* begin with `/**`.

From the [JSDoc getting started guide](https://jsdoc.app/about-getting-started.html):

> Each comment must start with a /** sequence in order to be recognized by the JSDoc parser. Comments beginning with /*, /***, or more than 3 stars will be ignored. This is a feature to allow you to suppress parsing of comment blocks.

**Comment by wojpawlik** [comment]
My proposal: deprecate `"removeComments"`, strip comments from output JS iff source maps are enabled.

**Comment by liukeyi300** [comment]
Why not just unify the behavior of ```removeComments```? If ```removeComments``` is false, preserve all comment in .d.ts and .js, otherwise strip all comment.

**Comment by Devorein** [comment]
There is a way to keep the comments in declaration files only and stripping them from the js files. But it's a little inefficient.

Basically, you'll need two tsconfig files.

1. tsconfig.json: Base tsconfig file
2. tsconfig.nocomments.json, which will have the following content

```json
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "declaration": false,
    "removeComments": true
  },
}
```
Now, while in development run the base tsconfig file, which will preserve the comments in .d.ts and .js files. But during production, run the base tsconfig file first, then the nocomments one. An example npm scripts could be:-

```json
"scripts": {
  "build": "tsc",
  "build:nocomments": "tsc -p ./tsconfig.nocomments.json",
  "prod": "npm run build && npm run build:nocomments"
},
```
Running `npm run prod`, will first generate .d.ts and .js files with comments, then it will generate only .js files without comments since `removeComments` is set to `true` and no .d.ts files since `declaration: false`, thus keeping the same commented .d.ts files.

All thanks to this original [post](https://dev.to/mudlabs/roll-your-own-comments-cleaner-3m79) in dev.to.

**Comment by equinusocio** [comment]
End 2021 and nothing on this... still embarrassing.

**Comment by lcoder** [comment]
hope this feature

**Comment by fantaclaus** [comment]
jsdoc @deprecated gets stripped off from d.ts files. but this is quite important thing.

**Comment by justinasfour04** [comment]
Is this going to be fixed

**Comment by LSafer** [comment]
So this problem became something big? 😪

I thought I could fix it in 5 minutes of googling

If you are here searching for an answer to this issue, DONT typescript does not support it!

**Comment by willster277** [comment]
This [suggestion](https://github.com/microsoft/TypeScript/labels/Suggestion) has been [in discussion](https://github.com/microsoft/TypeScript/labels/In%20Discussion) for ***over half a decade*** *(specifically: 2,041 days at the time of writing)*.

@RyanCavanaugh do you and your team plan to show an interest in features the community have actively wanted for 2041 days, or will you continue to ignore it for another ***five years***?

This feature is simple, highly requested, and is actually a bug report (`@deprecated` gets removed from dist code, **not good enough**). Many people, myself included, are getting increasingly tired of the TS team ignoring their users, we don't want TC39 to make `: string` a comment, we want this extremely small and simple request to be completed by someone, it will take mere minutes to do yourselves, or even less time to click "merge" on an [already submitted PR](https://github.com/microsoft/TypeScript/pull/31813) (this bug was fixed 3 years ago and is still not fixed in production, astounding!)

**Comment by webstrand** [comment]
@willster277 A workaround using project references: https://github.com/webstrand/typescript-remove-comments-not-from-declarations

Running `tsc --build` causes the `.js` files to be generated with `"removeComments": true` and the generation of `.d.ts` files is deferred to a sub-project which then generates them without `"removeComments": true`.

**Comment by willster277** [comment]
@webstrand it's shocking that we have to use 3rd party libraries to fix a Microsoft product.

Actually, it's exactly what I expect from Microsoft, which is arguably even worse.

**Comment by webstrand** [comment]
@willster277 Actually that's not 3rd-party. The repo's just an example of how to set the project up.

**Comment by willster277** [comment]
@webstrand my bad I'm on mobile, I thought you had sent a fork/project.

I have done this before, but it's not ideal. It would be much better if Typescript didn't destroy information.

**Comment by coreyward** [comment]
I use TypeScript to generate definitions for my JS source files (for use in libraries), and the lack of carryover from JSDoc comments on functions was disappointing. Since it doesn't seem this will be supported any time soon, I built a post-build script that copies over the JSDoc comments (less redundant type info) for functions declare in the `.d.ts` type files.

If anybody is interested in trying this approach, I've dropped the script in a gist here: https://gist.github.com/coreyward/e70642a41e2181641a817a632a82cd2d

**Comment by zzzgit** [comment]
``` json
removeComments: true,
preserveJSDoc: true
```

Hope it can be implemented like this.

**Comment by termermc** [comment]
One helpful thing would also be to remove type declarations within the preserved JSDoc comments as they're redundant, and may cause issues sometimes.

Additionally, sometimes I use `/** @typedef {import('./module.js').Type} Type */` to locally import types, and it's left in verbatim. Pseudo-imports like that should be turned into actual imports ideally.

**Comment by michal-kapala** [comment]
A sane workaround:
```
tsc --removeComments && tsc --emitDeclarationOnly
```

With `tsconfig.json` options:
```json
"removeComments": true,
"declaration": true,
```

**Comment by alexanderroidl** [comment]
This being wanted and not being implemented for about ten years is just sad. Classic Microsoft! 🎉🎉

**Comment by unphased** [comment]
> A sane workaround:
> 
> ```
> tsc --removeComments && tsc --declaration --emitDeclarationOnly
> ```
> 
> With `tsconfig.json` options:
> 
> ```json
> "removeComments": false,
> "declaration": false,
> ```
> 
> Builds stripped JS, then emits declarations with JSDoc preserved.

Could you explain this workaround? I have many projects already underway and I really need the d.ts to gain the comments. It seems like this solution requires a deep understanding of how the `tsconfig.json` options actually interact with command line flags, which I was hoping to avoid as the idea should be that anything you could specify in cli flags should also be configurable via config file...

**Comment by michal-kapala** [comment]
@unphased It is a simple 2-step build:
- the actual build
    - [`removeComments`](https://www.typescriptlang.org/tsconfig/#removeComments) strips all the comments from the code
    - declarations are not generated just yet (`declaration: false` in tsconfig)
- declaration emission
    - [`declaration`](https://www.typescriptlang.org/tsconfig/#declaration) tells `tsc` to generate `d.ts` files
    - [`emitDeclarationOnly`](https://www.typescriptlang.org/tsconfig/#emitDeclarationOnly) prevents javascript from rebuilding

Options passed via command line override the ones in `tsconfig.json`, there's not much more to this honestly

**Comment by unphased** [comment]
I figured it out. With my existing setup all I had to do was adjust my comments to Jason/javadoc style with `/**` and it works well.

**Comment by JounQin** [comment]
Oh, it's 2025 now.
